### ISSUE TITLE
When will Flask support native ASGI? Just like Django and FastAPI.

### ISSUE THREAD

**Issue by Cyberbolt** [pallets/flask#4661]
I don't want to refactor the code to Quart, but rather have the freedom to choose WSGI or ASGI when using Flask. Thank you for your reply!

**Comment by davidism** [comment]
Thank you for your interest. Just like questions about when we will make a new release, I don't plan to speculate on when any given feature may be available.

If you are invested in having ASGI available in Flask, the best ways for you to make this happen is to contribute, either code or funding, to the project.

Note that async has been available since day 1 of Flask, by using gevent or eventlet. ASGI is a modern take on WSGI for `async/await` but is in no way a requirement to get the performance of async workers.

We've also added async handling already, so that you can write `async def` views to take advantage of libraries that are built on `async/await` with less work. This does not provide the connection capabilities of ASGI, but does allow for async programming.

The only refactor that should be necessary to use Quart is to replace `Flask(__name__)` with `Quart(__name__)`. Any other incompatibilities with existing extensions will be equally true if Flask itself supports API, and requires continued updates throughout the ecosystem.

So in other words, this is an incredibly complex problem that we're already working on and continue to work on. Please contribute if you want it to come about sooner.

**Comment by davidism** [comment]
Yep, just reconfirming what I said above: this is an incredibly complex problem that we're already working on and continue to work on. Please contribute if you want it to come about sooner.

@miguelgrinberg for you specifically, since you have experience with AioFlask it would probably be really helpful for you to join the Flask and Werkzeug dev channels, and the monthly maintainer meeting, on our Discord. If you don't already have access, message me on https://discord.gg/pallets.

You're correct that it's likely impossible to have one class that does both sync and async with the same methods. We are _acutely_ aware of this, and have been discussing possibilities for years now. I just added @pgjones, author of Quart, as a core maintainer to Pallets, and we've been working for the last few years to identify how Werkzeug can be SansIO, and how Flask can have an ASGI interface while still supporting WSGI.

And you're also right that there are compatibility issues when attempting to use a Flask extension expecting WSGI in ASGI. We're also slowly setting up Pallets-Eco to be a home for community maintainership of extensions, so that we can maybe have a chance at updating them. But yeah, it's a huge amount of work beyond implementing ASGI itself, which is already a huge amount of work. This will be the case if Flask supports ASGI just like it's the case for Quart today. Encourage maintainers to transfer their projects to Pallets-Eco so they can remain maintained, and help them identify how they can be compatible with both.

For Flask-SQLAlchemy specifically, I've got a huge v3 refactor in the works. As part of that, I looked at async, and as far as I can tell there's no change required to support it. Pass the `async=True` option to the engine and session options config, and use an async driver, and it should just work. If there's something specific that is blocking in the extension itself, open a bug report about it or tell me about it on the dev channel on our Discord.

**Comment by miguelgrinberg** [comment]
> For Flask-SQLAlchemy specifically, I've got a huge v3 refactor in the works. As part of that, I looked at async, and as far as I can tell there's no change required to support it.

The main issue I had in mind is the fact that Mike Bayer decided to not port the `Query` object to asyncio, which means that it can only be used inside a `run_async` asyncio <--> greenlet bridges. This `Query` object is core to Flask-SQLAlchemy, so I don't see how it would work, unless run_async wrappers for its methods are created. SQLAlchemy 1.4 introduces a new querying interface that separates the building of a query (which is sync/async agnostic) from the execution of it, which has to go through different sync or async implementations of session, engine, etc. One of the issues with the `Query` class is that it mixes these two separate actions in ways that make them impossible to separate. For example, the `lazy='dynamic'` relationship mode could never work under asyncio.

**Comment by giuliohome** [comment]
Blocking me on Twitter simply because I referenced this issue as problematic makes me change my mind about Flask. Honestly I admired it and always talked very well about this open source inside the IT department of my company - till now. After noticing you have blocked me for what I have written on Twitter, I 'll definitely stay away from this project and I'll discourage anyone to rely on this software implementation. Since I used it in the past for an application in production, my users will have to bear the transaction time and effort to a different web app server. You write that the best way to implement async etc...is funding this project, in a way that, honestly, looks arrogant. My reaction is that the best way to implement something is starting from the current state of the art, without reinventing the wheel but leveraging those open source or otherwise those providers who already show a meaningful knowledge basis to really enable the implementation.

**Comment by pgjones** [comment]
@giuliohome I'm going to hide your comment and this one as they will both be off topic. I'm unsure about the twitter specifics so I won't comment there. We are making a great deal of progress towards ASGI, but there are large technical issues that prevent it happening. Timing to work on these projects certainly does help, I've made more progress this last month (having been made redundant) than I have all year. Funding helps with this as we can buy time for people to solve the problems.

I would summarise that Flask can easily support native ASGI if compatibility with existing code, and crucially extensions is broken (or large scale monkeypatching is acceptable). This has been done already, and I called it [Quart](https://github.com/pgjones/quart). As we don't want to do this we'll keep working towards another solution, for example [async support](https://flask.palletsprojects.com/en/2.1.x/async-await/) in Flask itself.

**Comment by davidism** [comment]
Locking this preemptively as unfortunately it seems to have been linked from elsewhere in a way that will get off topic.
